sips/sip-021/sip-021-trustless-two-way-peg-for-bitcoin.md

```diff
@@ -0,0 +1,1385 @@
+
+Marvin Janssen (of Ryder) has a proposal for a "Bitcoin train" taproot script
+for recovering from a peg liveness failure that should be given further
+consideration.
+
+# Backwards Compatibility
+
+Two properties from the Stacks blockchain are dropped in this proposal:
+
+* **The ability to mine off of any prior Stacks block.**  To date, there has
+never been a Stacks block whose ancestor was over 100 Bitcoin blocks back (or
+even 10 Bitcoin blocks back).  Every deep fork in Stacks that has ever arisen
+was due to a bug, not miners leveraging this feature.  Furthermore, the
+necessity of this feature is questionable, since its need is motivated by an
+already-catastrophic blockchain failure (in which case, other remedies such as
+hard forks would be viable alternatives).
+
+* **The ability for miners to choose their block-commit value on-the-fly.** In
+this proposal, miners must choose a block-commit value for the next block they
+intend to mine.  However, to date miners do not appear to dynamically adjust
+their PoX expenditures every block as it is, so this feature's ongoing utility
+is questionable.
+
+# Activation
+
+This section will be filled in with the advice of the CABs.
+
+# Conclusion
+
+This proposal describes sBTC, a novel Bitcoin 2-way-peg that offers a
+PoX-powered recovery procedure for temporary peg liveness failures while
+rewarding the peg wallet's stewards (Stackers, via PoX) for staying online
+regardless of how much or little peg traffic exists.  The proposal sacrifices
+forks of arbitrary length in order to handle peg-ins and peg-outs of arbitrary
+size in a fixed amount of time in an in-band fashion, while keeping mining and
+Stacking running in an open-membership setting.  Building the peg on Stacks
+avoids complicated designs related to Bitcoin forking by ensuring that the peg
+state exists in all Bitcoin forks.  By making sure that Stacking and
+peg-operations materialize in all Stacks forks, we additionally ensure that
+mining on the canonical chain is always the most profitable course of action
+for miners.
+
+# Appendix A: New Transactions
+
+This section describes the wire formats of the various sBTC Bitcoin
+transactions.
+
+### Peg Hand-off Transaction
+
+* **Description**: This is used by the current reward cycle's Stackers to send
+the pegged-in BTC to the next reward cycle's Stackers' peg wallet address.
+
+* **Opcode**: `H`
+
+* **Data**
+
+	* **Bytes 3-11**: The reward cycle number
+(big-endian) corresponding to the Stacker set doing the hand-off
+
+* **Output 2**: the new peg wallet address
+
+* **Notes**: Anyone can send a hand-off
+transaction in order to (re-)fill the current reward cycle's peg wallet.  This
+is used to recover from loss of BTC due to transaction fees, or even wallet
+theft.
+
+
+### Peg-In Transaction
+
+* **Description**: This is used by an sBTC user to peg-in BTC and materialize
+sBTC. 
+
+* **Opcode**: `<`
+
+* **Data**
+
+	* **Bytes 3-80**: c32-encoded Stacks
```

Comment:
This should be _any_ valid Stacks address (principal or smart contract), and moreover, the address part does not need to be c32-encoded.  Encoding it as a 21-byte string is fine.

Reply:
Do you mean it should be the hex-encoded principal, using Clarity serialization?

Reply:
I mean:

* byte 3: address version
* bytes 4-23: address hash bytes
* bytes 24-64: address name, if this is a smart contract